### ISSUE TITLE
Revert back to limiting number of threads when not explicitely provided

### ISSUE THREAD

**Issue by NewbiZ** [pydata/numexpr#328]
The latest `numexpr` patch release (2.6.8 -> 2.6.9) had a change that had a much more important impact than what the changelog describes:
> more robust handling of the thread-setting environment variables

There were an upper bound on the number of threads used before (8) which is not here anymore. A process that was using 8 threads previously can now use up to `<machine CPU count>` threads now! This is even more dramatic when using multiprocessing. If you previously had 8 processes using 8 threads (for a total of 64 threads, which was perfectly fine on a 64 cores machine), with 2.6.9 it is now using 8*64=512 threads.

Considering that `numexpr` is 99% of the time an indirect dependency (from pandas, numpy, etc), which do not enforce patch-level dependency, such high impact changes are potential production-breaking (this happened to us today, for this exact reason).

I don't think convincing other projects to depend on `numexpr` up to patch level is reasonable, so `numexpr` should really try to not have impacting changes without changing at least the minor version.

**Comment by NewbiZ** [comment]
I would also suggest to add back the limit of 8 threads. I routinely do (and know a lot of other people in the same situation) run `numexpr`/`numpy` computations on machines with 128 cores. I don't think it's sane to let `numexpr` spawn 128 threads _out of the box_ when no environment variable is configured.

**Comment by FrancescAlted** [comment]
Agreed, I am +1 on the suggestion to add back the limit of 8 threads.

**Comment by gdementen** [comment]
Disclaimer: I am not concerned by this issue (I don't have access to those nice servers). 

I just looked at the code for curiosity and I wonder if the more sensible option wouldn't be to cap at 8 (or whatever number) only in the case where it was set via detect_number_of_cores() and not in the case where it was set via OMP_NUM_THREADS. It was capped in both cases until 2.6.8 and in 2.6.9, it is never capped.

FWIW, the old code was:
```python
    try:
        nthreads = int(os.environ['NUMEXPR_NUM_THREADS'])
    except KeyError:
        nthreads = int(os.environ.get('OMP_NUM_THREADS', detect_number_of_cores()))
        # Check that we don't activate too many threads at the same time.
        # 8 seems a sensible value.
        if nthreads > 8:
            nthreads = 8
```

**Comment by oliparcol** [comment]
> I would also suggest to add back the limit of 8 threads. 

I strongly agree, it doesn't seem safe to use all cores of a server.

**Comment by pmuller** [comment]
It would be great to revert to the previous default. Here we share machines across jobs, and we ended up with weird errors (e.g. suddent lack of resources, issues when spawning threads, etc.) before tracking down this minor change. The only way this change makes sense is for people who _only_ run jobs in containers or dedicated servers, which I do not believe is the majority of numexpr users.

**Comment by robbmcleod** [comment]
Probably 2.6.5 should have been a minor version bump, since that fixed the trouble with the massive 4096 thread threadpool. I guess in my mind I have had NumExpr 2.6 in maintenance mode.

When using "hybrid MPL"-style mixture of multiprocessing and multithreading in a cluster environment, if you aren't controlling the number of threads per process, that's a logical bug. Cluster scheduling system don't force core affinity (that I have ever seen), so it's generally an honour system that you use the number of cores that you request from the queuing system. In the past version, if you launched 16 processes on a 64-physical core node, you would likely just get away with using 128 threads total thanks to hyperthreading. With the unbound use of cores, your cluster administrators are going to notice in their logs that you have massively oversubscribed a node and thus haven't configured your job correctly. 

Apparently previously the default maximum was covering up this error. So we could revert it back to capping auto-detection at 8 cores. Or we could disable auto-detection of cores completely and use a default of 1 core unless the appropriate environment variables are set, or the developer calls `numexpr.set_num_threads()`. That would absolutely force people to actually read the documentation and understand how to configure the software, without breaking any systems.

Alternatively, if we do revert to 8 threads, I would prefer to add a warning if they do have more than 8 virtual cores, so people notice that they haven't configured it.

It would also probably be a good idea to add to our documentation a brief tutorial on how to setup NumExpr to work in a cluster environment.

**Comment by NewbiZ** [comment]
> in a cluster environment, if you aren't controlling the number of threads per process, that's a logical bug.

The issue is that not everyone use `numexpr` directly, it just happens to be a `n`th level dependency in the stack. In these (common) cases, it's not really reasonable to assume the top level component/maintainer has knowledge of what is, and how to configure `numexpr`.
For instance in my case, the dependency tree leading to `numexpr` is: `pandas` -> `tables` -> `numexpr`. TBH I did not even know what was `numexpr` before everything started exploding with `2.6.9`.

>  Cluster scheduling system don't force core affinity (that I have ever seen), so it's generally an honour system that you use the number of cores that you request from the queuing system.

It's not even really an issue of stealing other's resources IMHO (btw, e.g. `slurm` does force affinity through `cgroups`). On heavy computing nodes, it could just be a matter of system configuration (i.e. `PTHREAD_THREADS_MAX==1024`)

> Or we could disable auto-detection of cores completely and use a default of 1 core unless the appropriate environment variables are set, or the developer calls numexpr.set_num_threads(). That would absolutely force people to actually read the documentation and understand how to configure the software, without breaking any systems.

I also like that solution. It's simple and conservative.

> Alternatively, if we do revert to 8 threads, I would prefer to add a warning if they do have more than 8 virtual cores, so people notice that they haven't configured it.

I can add that warning in the PR if you want.

**Comment by robbmcleod** [comment]
For anyone waiting on this I haven't forgotten, I'm just crunching at work.

**Comment by robbmcleod** [comment]
Take a look at the proposed fix in 7c2ef387d81cd450e8220fe4174cf46ec559994c. I decided that I needed to write a new function to make sure that I got the logic correct. I also moved the platform specific check for Sparc out of `__init__.py` as it really doesn't belong there.

Also look at the changes in the user guide and let me know if there's anything not clear there.

If anyone can think of other regression tests please let me know. I added a couple for `NUMEXPR_MAX_THREADS` launching via `subprocess`. Unfortunately `importlib.reload()` doesn't work for testing environment variables as it doesn't really restart the C-extension.

**Comment by NewbiZ** [comment]
Changes looks fine to me, and the doc is very clear. Thanks a lot @robbmcleod 
I think we can close the issue :)

**Comment by robbmcleod** [comment]
I'll leave this one open thanks, until we do a release, just to keep the visibility high.

**Comment by robbmcleod** [comment]
Closing as was resolved with the release of `2.7.0`.
